From 46478fec31a4a310131a33c0422da32bbb3b445e Mon Sep 17 00:00:00 2001
From: Mingyu Lei <leimy@chromium.org>
Date: Fri, 01 Nov 2024 08:09:27 +0000
Subject: [PATCH] Prompt API: stop storing ReceiverContext for EchoAIManagerImpl

The receiver context is not used in EchoAIManagerImpl, and it may lead
to use-after-free issues. This CL removes it.

Bug: 376667975
Change-Id: Ice872c00c91d9452d70b505508e7f4f48815f056
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5981660
Commit-Queue: Mingyu Lei <leimy@chromium.org>
Reviewed-by: Rakina Zata Amni <rakina@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1376787}
---

--- a/content/browser/ai/echo_ai_manager_impl.cc
+++ b/content/browser/ai/echo_ai_manager_impl.cc
@@ -24,10 +24,10 @@
 
 // static
 void EchoAIManagerImpl::Create(
-    ReceiverContext context,
+    std::variant<RenderFrameHost*, base::SupportsUserData*> context,
     mojo::PendingReceiver<blink::mojom::AIManager> receiver) {
   static base::NoDestructor<EchoAIManagerImpl> ai;
-  ai->receivers_.Add(ai.get(), std::move(receiver), context);
+  ai->receivers_.Add(ai.get(), std::move(receiver));
 }
 
 void EchoAIManagerImpl::CanCreateAssistant(
--- a/content/browser/ai/echo_ai_manager_impl.h
+++ b/content/browser/ai/echo_ai_manager_impl.h
@@ -22,16 +22,14 @@
 // will be set using the default value.
 class EchoAIManagerImpl : public blink::mojom::AIManager {
  public:
-  using ReceiverContext =
-      std::variant<RenderFrameHost*, base::SupportsUserData*>;
-
   EchoAIManagerImpl(const EchoAIManagerImpl&) = delete;
   EchoAIManagerImpl& operator=(const EchoAIManagerImpl&) = delete;
 
   ~EchoAIManagerImpl() override;
 
-  static void Create(ReceiverContext context,
-                     mojo::PendingReceiver<blink::mojom::AIManager> receiver);
+  static void Create(
+      std::variant<RenderFrameHost*, base::SupportsUserData*> context,
+      mojo::PendingReceiver<blink::mojom::AIManager> receiver);
 
  private:
   friend base::NoDestructor<EchoAIManagerImpl>;
@@ -59,7 +57,7 @@
       mojo::PendingRemote<blink::mojom::AIManagerCreateRewriterClient> client,
       blink::mojom::AIRewriterCreateOptionsPtr options) override;
 
-  mojo::ReceiverSet<blink::mojom::AIManager, ReceiverContext> receivers_;
+  mojo::ReceiverSet<blink::mojom::AIManager> receivers_;
 };
 
 }  // namespace content
